src/directives/NgGrid.ts

```diff
@@ -902,108 +941,103 @@ export class NgGrid implements OnInit, DoCheck, OnDestroy {
 			dims = this._resizingItem.getSize();
 		}
 
+		let itemsInGrid: NgGridItem[] = Array.from(this._itemsInGrid.values());
+
 		switch (this.cascade) {
 			case 'up':
 			case 'down':
+				itemsInGrid = itemsInGrid.sort(NgGrid.sortItemsByPositionVertical);
 				const lowestRowPerColumn: Map<number, number> = new Map<number, number>();
 
-				for (let row: number = 1; row <= this._curMaxRow; row++) {
-					if (this._itemGrid[row] == undefined) continue;
+				for (let item of itemsInGrid) {
+					if (item.isFixed) continue;
 
-					for (let col: number = 1; col <= this._curMaxCol; col++) {
-						if (this._itemGrid[row] == undefined) break;
-						if (lowestRowPerColumn.has(col) && row < lowestRowPerColumn.get(col)) continue;
+					const itemDims: NgGridItemSize = item.getSize();
+					const itemPos: NgGridItemPosition = item.getGridPosition();
 
-						if (this._itemGrid[row][col] != null) {
-							const item: NgGridItem = this._itemGrid[row][col];
-							if (item.isFixed) continue;
+					let lowestRowForItem: number = lowestRowPerColumn.has(itemPos.col) ? lowestRowPerColumn.get(itemPos.col) : 1;
 
-							const itemDims: NgGridItemSize = item.getSize();
-							const itemPos: NgGridItemPosition = item.getGridPosition();
+					for (let i: number = 1; i < itemDims.x; i++) {
+						const lowestRowForColumn = lowestRowPerColumn.has(itemPos.col + i) ? lowestRowPerColumn.get(itemPos.col + i) : 1;

```

Comment:
I thinik that "const lowestRowForColumn = lowestRowPerColumn.get(itemPos.col + i) || 1;" can be used too.